File: src/server/routes/actions.ts
import type { FastifyInstance } from "fastify";
import type { PendingActionStore } from "../../pending-action/index.js";
import type { WebSocketChannel } from "../../pending-action/ws-channel.js";
import { toActionView } from "../../pending-action/index.js";
import { isUserVerified } from "../../webauthn/signature-format.js";

export interface ActionRoutesParams {
  app: FastifyInstance;
  actionStore: PendingActionStore;
  wsChannel: WebSocketChannel;
}

export function registerActionRoutes(params: ActionRoutesParams) {
  const { app, actionStore, wsChannel } = params;

  app.get<{ Params: { actionId: string } }>("/api/actions/:actionId", async (request, reply) => {
    if (!request.accountId) {
      reply.status(401);
      return { error: "Not authenticated" };
    }

    const action = actionStore.get(request.params.actionId);
    if (!action) {
      reply.status(404);
      return { error: "Action not found" };
    }
    if (action.accountId !== request.accountId) {
      reply.status(403);
      return { error: "Access denied" };
    }

    return toActionView(action);
  });

  app.post<{
    Params: { actionId: string };
    Body: { authenticatorData: string; signature: string; clientDataJSON: string };
  }>("/api/actions/:actionId/resolve", async (request, reply) => {
    if (!request.accountId) {
      reply.status(401);
      return { error: "Not authenticated" };
    }

    const action = actionStore.get(request.params.actionId);
    if (!action) {
      reply.status(404);
      return { error: "Action not found" };
    }
    if (action.accountId !== request.accountId) {
      reply.status(403);
      return { error: "Access denied" };
    }
    if (action.status !== "pending") {
      reply.status(409);
      return { error: `Action is already ${action.status}` };
    }

    let resolved: boolean;

    if (action.type === "webauthn-sign") {
      const { authenticatorData, signature, clientDataJSON } = request.body ?? {};
      if (
        typeof authenticatorData !== "string" ||
        typeof signature !== "string" ||
        typeof clientDataJSON !== "string"
      ) {
        reply.status(400);
        return { error: "Missing required fields: authenticatorData, signature, clientDataJSON" };
      }
      const authDataBuf = Buffer.from(authenticatorData, "base64url");
      // Defense-in-depth: when the action (typically the originating endpoint)
      // requires UV, reject responses without the UV bit set so we fail fast
      // with a clear error. Cryptographic enforcement still happens downstream
      // when the SSH server verifies the signature against the sk-* pubkey and
      // its `verify-required` option; this check alone is not a primary gate.
      // For "preferred"/"discouraged" endpoints we accept either UV state.
      if (action.userVerification === "required" && !isUserVerified(authDataBuf)) {
        reply.status(400);
        return { error: "User verification required" };
      }
      resolved = actionStore.resolve(action.id, {
        requestId: action.id,
        authenticatorData: authDataBuf,
        signature: Buffer.from(signature, "base64url"),
        clientDataJSON,
      });
    } else {
      // key-approve: no payload needed
      resolved = actionStore.resolve(action.id);
    }

    if (!resolved) {
      reply.status(409);
      return { error: "Action could not be resolved" };
    }

    wsChannel.broadcastResolved(action.id, action.accountId);
    return { redirectTo: action.redirectTo };
  });

  app.post<{ Params: { actionId: string } }>(
    "/api/actions/:actionId/deny",
    async (request, reply) => {
      if (!request.accountId) {
        reply.status(401);
        return { error: "Not authenticated" };
      }

      const action = actionStore.get(request.params.actionId);
      if (!action) {
        reply.status(404);
        return { error: "Action not found" };
      }
      if (action.accountId !== request.accountId) {
        reply.status(403);
        return { error: "Access denied" };
      }
      if (action.status !== "pending") {
        reply.status(409);
        return { error: `Action is already ${action.status}` };
      }

      actionStore.deny(action.id);
      wsChannel.broadcastResolved(action.id, action.accountId);
      return { status: "denied" };
    },
  );
}

File: internal/rest/actions.go
// SPDX-License-Identifier: LicenseRef-FSL-1.1-Apache-2.0
// Pending-action REST (port of src/server/routes/actions.ts): the /sign/:id
// page reads GET /api/actions/:id, then POSTs the WebAuthn assertion to
// resolve (or empty to deny). Account-scoped; state conflicts 409.
package rest

import (
	"encoding/base64"
	"encoding/json"
	"net/http"

	"github.com/go-chi/chi/v5"

	"github.com/rado0x54/shellwatch/internal/approval"
	"github.com/rado0x54/shellwatch/internal/signing"
)

// Actions wires the pending-action routes.
type Actions struct {
	Store *approval.Store
}

func (a *Actions) Mount(r chi.Router) {
	r.Get("/api/actions/{actionId}", a.get)
	r.Post("/api/actions/{actionId}/resolve", a.resolve)
	r.Post("/api/actions/{actionId}/deny", a.deny)
}

func (a *Actions) lookup(w http.ResponseWriter, r *http.Request) *approval.Action {
	action := a.Store.Get(chi.URLParam(r, "actionId"))
	if action == nil {
		writeErr(w, 404, "Action not found")
		return nil
	}
	if action.AccountID != accountID(r) {
		writeErr(w, 403, "Access denied")
		return nil
	}
	return action
}

func (a *Actions) get(w http.ResponseWriter, r *http.Request) {
	action := a.lookup(w, r)
	if action == nil {
		return
	}
	writeJSON(w, 200, actionView(action))
}

func (a *Actions) resolve(w http.ResponseWriter, r *http.Request) {
	action := a.lookup(w, r)
	if action == nil {
		return
	}
	if action.Status != approval.StatusPending {
		writeErr(w, 409, "Action is already "+string(action.Status))
		return
	}
	switch action.Type {
	case approval.TypeWebAuthnSign:
		body := readRawBody(r)
		authData := b64urlField(body, "authenticatorData")
		sig := b64urlField(body, "signature")
		cdj := b64urlField(body, "clientDataJSON")
		if authData == nil || sig == nil || cdj == nil {
			writeErr(w, 400, "Missing required fields: authenticatorData, signature, clientDataJSON")
			return
		}
		// Defense-in-depth UV check (the SSH server is the real gate).
		if action.UserVerification == "required" && !signing.IsUserVerified(authData) {
			writeErr(w, 400, "User verification required")
			return
		}
		if !a.Store.ResolveSign(action.ID, signing.SignResponse{
			AuthenticatorData: authData, Signature: sig, ClientDataJSON: cdj,
		}) {
			writeErr(w, 409, "Action is already "+string(action.Status))
			return
		}
	case approval.TypeKeyApprove:
		if !a.Store.ResolveKey(action.ID) {
			writeErr(w, 409, "Action is already "+string(action.Status))
			return
		}
	}
	writeJSON(w, 200, map[string]any{"status": "resolved", "redirectTo": action.RedirectTo})
}

func (a *Actions) deny(w http.ResponseWriter, r *http.Request) {
	action := a.lookup(w, r)
	if action == nil {
		return
	}
	if action.Status != approval.StatusPending {
		writeErr(w, 409, "Action is already "+string(action.Status))
		return
	}
	a.Store.Deny(action.ID)
	writeJSON(w, 200, map[string]any{"status": "denied"})
}

// actionView is the client-safe projection (excludes resolve/reject closures).
func actionView(a *approval.Action) map[string]any {
	v := map[string]any{
		"id": a.ID, "accountId": a.AccountID, "type": string(a.Type), "status": string(a.Status),
		"createdAt": a.CreatedAt.UTC().Format(isoMillis), "expiresAt": a.ExpiresAt.UTC().Format(isoMillis),
		"context": a.Context,
	}
	if a.RedirectTo != "" {
		v["redirectTo"] = a.RedirectTo
	}
	switch a.Type {
	case approval.TypeWebAuthnSign:
		v["credentialId"] = a.CredentialID
		v["challenge"] = a.Challenge
		v["rpId"] = a.RpID
		v["userVerification"] = a.UserVerification
		if a.PasskeyLabel != "" {
			v["passkeyLabel"] = a.PasskeyLabel
		}
	case approval.TypeKeyApprove:
		v["keyLabel"] = a.KeyLabel
		v["keyFingerprint"] = a.KeyFingerprint
	}
	return v
}

func b64urlField(body map[string]json.RawMessage, key string) []byte {
	raw, ok := body[key]
	if !ok {
		return nil
	}
	s := jsonString(raw)
	if s == "" {
		return nil
	}
	d, err := base64.RawURLEncoding.DecodeString(s)
	if err != nil {
		return nil
	}
	return d
}
